### PR TITLE
stack nd array w/unknown chunks

### LINKED ISSUE #7546: numpy/dask difference in stacking multi-dim arrays with unknown chunk sizes
**What happened**:

When using da.stack and stacking a multi-dim list (of arrays), the allow_unknown_chunksizes parameter is not passed into asarray.  This is needed as asarray calls back to stack if it is passed a sequence, and an error is raised if the unknown chunksize is not the first dimension.

**What you expected to happen**:

da.stack and np.stack to produce the same result when stacking a multi-dim array.

**Minimal Complete Verifiable Example**:

```python
import numpy as np
import pandas as pd
import dask.array as da
import dask.dataframe as dd

a_df = pd.DataFrame({"a": np.arange(12)})
b_df = pd.DataFrame({"b": np.arange(12) * 10})
c_df = pd.DataFrame({"c": np.arange(12) * 100})
d_df = pd.DataFrame({"d": np.arange(12) * 1000})

a_ddf = dd.from_pandas(a_df, sort=False, npartitions=3)
b_ddf = dd.from_pandas(b_df, sort=False, npartitions=3)
c_ddf = dd.from_pandas(c_df, sort=False, npartitions=3)
d_ddf = dd.from_pandas(d_df, sort=False, npartitions=3)

a_x = a_ddf.values
b_x = b_ddf.values
c_x = c_ddf.values
d_x = d_ddf.values

# Works
np.stack([
    [a_df.values, b_df.values],
    [c_df.values, d_df.values],
])

# Does not work
da.stack([
    [a_x, b_x],
    [c_x, d_x]
], allow_unknown_chunksizes=True)
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2021.04.0+16.ge83379d5
- Python version: 3.8.8
- Operating System: MacOS 11.2.3
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by chrisroat**
- [x] Closes #7546
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -4122,7 +4122,7 @@ def asarray(a, **kwargs):
     elif type(a).__module__.split(".")[0] == "xarray" and hasattr(a, "data"):
         return asarray(a.data)
     elif isinstance(a, (list, tuple)) and any(isinstance(i, Array) for i in a):
-        return stack(a)
+        return stack(a, allow_unknown_chunksizes=allow_unknown_chunksizes, **kwargs)
```
Comment (jsignell) [inline dask/array/core.py:4125]: I don't think we should be passing the `kwargs` into this stack call.

```suggestion
        return stack(a, allow_unknown_chunksizes=allow_unknown_chunksizes)
```

**Review by jsignell — COMMENTED** [review]
This looks good, I just have one small comment. And sorry this sat for so long :(

File: dask/array/core.py (hunk 1 above)
Comment (chrisroat) [reply dask/array/core.py:4151]: I agree.  Thanks.

**Comment by chrisroat** [comment]
This was the first time I've used GitHub's UI to accept a change.  It looks like it didn't work as intended.  I will revert the latest change and force-push an updated change on my fork.

**Comment by chrisroat** [comment]
I note that `concatenate` does not pass in the `allow_unknown_chunksizes` parameter, so I added that.  There are already tests for this situation, and I would need to understand better why they might not need this.

I also note that there are 2 other related changes that could come up in the future:
- broadcast_to could possibly use allow_unknown_chunksizes, but that would require some care
- within `asarray`, the xarray fork could pass in allow_unknown_chunksizes to recursive `asarray` call.  However, xarray currently doesn't support unknown chunk sizes, so this is moot at the moment.

**Comment by jsignell** [comment]
This looks good to me, thanks for the work @chrisroat!  Note @quasiben @rjzamora I don't think this has cudf implications, but I'm happy to make changes if it does.

**Comment by quasiben** [comment]
Thanks for the ping @jsignell .  I tested `test_cupy`, `test_array_function.py`, and `test_random.py`  -- all passed!

**Pull request merged**